boot/bootutil/include/bootutil/image.h

```diff
@@ -82,16 +87,24 @@ struct image_version {
     uint32_t iv_build_num;
 };
 
+struct image_dependency {
```

Comment:
I think we need to solve these problems as well:
 - What if an update introduces breaking changes to an image that other images depend on?
 - What if the incompatible `foo` and `bar` images could both be found under the same `image_id` inside the user's ecosystem?

Possible solutions:
 - Add a Software ID requirement to this struct, that must match an ID in the TLV of the dependency image. This can become complex, especially if updates are allowed to change the ID of an image, but is also the most flexible.
 - Have a maximum version, either:
   - Add an `image_max_version`, or
   - Have an implicit max version, e.g. an image must have the same major version as the requirement.

Reply:
> What if the incompatible foo and bar images could both be found under the same image_id inside the user's ecosystem?

Basically, the `image_id` here is just an index of the image areas in the flash. It refers to one specific flash area.

Reply:
> What if an update introduces breaking changes to an image that other images depend on?

Could you clarify this situation, please? What update strategy is used?

Reply:
> Could you clarify this situation, please?

Imagine the following: The application (image0) depends on library `foo` (image1). `foo` follows semantic versioning, which means that all versions `n.a.b` are backwards compatible with older versions `n.c.d` as long as the major version `n` remains unchanged. However, if the major version changes, `foo` cannot be guaranteed to be compatible with the application any more. With the current `image_dependency` struct, there is no way to express this fact. So if `foo` is upgradeable without upgrading the app, it might be upgraded to an incompatible version.

Reply:
> Basically, the image_id here is just an index of the image areas in the flash. It refers to one specific flash area.

Exactly. We need something extra, that helps uniquely identify a particular library.

Reply:
> * What if the incompatible `foo` and `bar` images could both be found under the same `image_id` inside the user's ecosystem?

Image_id reflects to the image area in the flash. It is assumed that the minimum update-able unit is one flash area. If that are contains more than one executable then it is the user responsibility to assure their compatibility. They can be updated only together.

Reply:
> If that are contains more than one executable

Sorry, I meant if one or the other could be placed in the same image area. For example, if a future update re-purposes the area for another library. Or the user has two classes of devices, which place different libraries in that image area.

Reply:
> Imagine the following: The application (image0) depends on library `foo` (image1). `foo` follows semantic versioning, which means that all versions `n.a.b` are backwards compatible with older versions `n.c.d` as long as the major version `n` remains unchanged. However, if the major version changes, `foo` cannot be guaranteed to be compatible with the application any more. With the current `image_dependency` struct, there is no way to express this fact. So if `foo` is upgradeable without upgrading the app, it might be upgraded to an incompatible version.

Let say these are compatible
app 1.0.0 - 1.x.x  -  foo 2.0.0 - 2.x.x
app 2.0.0 - 2.x.x  -  foo 3.0.0 - 3-x.x.

In this case user must append a dependency TLV to foo newest (already incompatible) version of foo (3.x.x) to mandate a newer version of the app (at least 2.x.x) . So the user must be careful to attach all the required minimum dependencies to a new image which is not backward compatible. 

Other thing is that user do not use semantic versioning then you might cannot specify the maximum version in  advanced.

Reply:
It is interesting to note that, at least currently, SUIT dismisses entirely with versions and any complexity around this in regards to dependencies. The version field is a simple unsigned integer (can be arbitrarily large, though) and is just compared for upgrades. For dependencies, they just specify an exact image that they depend upon.

I'm not sure the bootloader is really the place to try and have meaningful semantic versioning, especially since we don't have enough information in our version structure to represent a proper semantic version.

The code to compare semantic versions is fairly complex, and is certainly not what our current `boot_is_version_sufficient` function is doing (I'll comment on the function itself).

If we do really think there are requirements for richer dependency information, we should probably engage with SUIT so that also gets represented in that manifest.

Reply:
> In this case user must append a dependency TLV to foo newest (already incompatible) version of foo (3.x.x) to mandate a newer version of the app (at least 2.x.x) . So the user must be careful to attach all the required minimum dependencies to a new image which is not backward compatible.

You're right, this does solve the problem. I see a few issues, though.

As a general observation: The "backwards" dependency from foo on the app is not an accurate expression of the relationship between foo and the app, since strictly speaking foo works correctly regardless of the app. Basically, it complicates the dependency tree (cyclic dependencies).

This will cause problems whenever you want to separate responsibility, i.e. when you have multiple FW owners. The maintainer of foo doesn't want to have to sign individual packages for each user of foo. Instead they want to create a single package for a given version, and sign it once.

> Other thing is that user do not use semantic versioning then you might cannot specify the maximum version in advanced.

This is only partially true IMO, just look at semantic versioning, if a new version of foo introduces breaking changes, the major version must be bumped. This is known beforehand by the app.

Reply:
> It is interesting to note that, at least currently, SUIT dismisses entirely with versions and any complexity around this in regards to dependencies. The version field is a simple unsigned integer (can be arbitrarily large, though) and is just compared for upgrades. For dependencies, they just specify an exact image that they depend upon.

Yes, MCUBoot is ahead of SUIT here.

> I'm not sure the bootloader is really the place to try and have meaningful semantic versioning, 

We need to set a limit somewhere with regards to complexity, I totally agree. I think that this patch, by introducing such rich a version scheme (major+minor+revision+build), should go further, in order to make the feature more complete. Using semver is one way, adding a max_version is another more flexible variant.

> especially since we don't have enough information in our version structure to represent a proper semantic version.

Right, there's no prerelease or build information, but we could have a more complete subset. We do have enough info to support semver with no pre-release versions (hyphens) or build metadata (plus signs)

> The code to compare semantic versions is fairly complex, and is certainly not what our current boot_is_version_sufficient function is doing (I'll comment on the function itself).

Adding the support for the above subset should not be hard, neither should supporting a max version. https://gist.github.com/oyvindronningstad/601ef3e1ef539db579b2316fc567852f

> If we do really think there are requirements for richer dependency information, we should probably engage with SUIT so that also gets represented in that manifest.

Definitely.